hazelcast/invocation.py

```diff
@@ -1,141 +1,175 @@
+    EXCEPTION_MESSAGE_TYPE
 from hazelcast.future import Future
-from hazelcast.lifecycle import LIFECYCLE_STATE_CONNECTED
-from hazelcast.protocol.client_message import LISTENER_FLAG
-from hazelcast.protocol.custom_codec import EXCEPTION_MESSAGE_TYPE, ErrorCodec
 from hazelcast.util import AtomicInteger
-from hazelcast.six.moves import queue
 from hazelcast import six
 
 
-class Invocation(object):
-    sent_connection = None
-    timer = None
-
-    def __init__(self, invocation_service, request, partition_id=-1, address=None, connection=None, event_handler=None):
-        self._event = threading.Event()
-        self._invocation_timeout = invocation_service.invocation_timeout
-        self.timeout = self._invocation_timeout + time.time()
-        self.address = address
-        self.connection = connection
-        self.partition_id = partition_id
-        self.request = request
-        self.future = Future()
-        self.event_handler = event_handler
+def _no_op_response_handler(_):
+    pass
 
-    def has_connection(self):
-        return self.connection is not None
 
-    def has_partition_id(self):
-        return self.partition_id >= 0
+class Invocation(object):
+    __slots__ = ("request", "timeout", "partition_id", "uuid", "connection", "event_handler",
+                 "future", "sent_connection", "timer", "urgent", "response_handler")
 
-    def has_address(self):
-        return self.address is not None
+    def __init__(self, request, partition_id=-1, uuid=None, connection=None,
+                 event_handler=None, urgent=False, timeout=None, response_handler=_no_op_response_handler):
+        self.request = request
+        self.partition_id = partition_id
+        self.uuid = uuid
+        self.connection = connection
+        self.event_handler = event_handler
+        self.urgent = urgent
+        self.timeout = timeout
+        self.future = Future()
+        self.timeout = None
+        self.sent_connection = None
+        self.timer = None
+        self.response_handler = response_handler
 
     def set_response(self, response):
         if self.timer:
             self.timer.cancel()
-        self.future.set_result(response)
+        try:
+            result = self.response_handler(response)
+            self.future.set_result(result)
+        except Exception as e:
+            self.future.set_exception(e)
 
     def set_exception(self, exception, traceback=None):
         if self.timer:
             self.timer.cancel()
         self.future.set_exception(exception, traceback)
 
-    def set_timeout(self, timeout):
-        self._invocation_timeout = timeout
-        self.timeout = self._invocation_timeout + time.time()
-
     def on_timeout(self):
-        self.set_exception(TimeoutError("Request timed out after %d seconds." % self._invocation_timeout))
+        self.set_exception(HazelcastTimeoutError("Request timed out."))
```

Comment:
This is different than the java side behavior. We are not cancelling the invocation when invocation.timeout.seconds reached.

Reply:
Good catch. Right now, client removes the pending connection on timeout. Also, added a test that verifies that there is nothing on the pending list after timeout

Reply:
I have actually meant that java client is not doing anything when invocation timeout is reached. 
It could be the case that invocation is a long-running task. We should check the timeout only when an exception occurs. Otherwise, we should not take any action.

Reply:
Oh, I thought that property means what I did since we had a timer for the timeout before in the Python client. It was added 5 years ago (https://github.com/hazelcast/hazelcast-python-client/commit/d962b8820239af8d24fc8db6e3eef7cda717b2e4) maybe it meant something like this back then :smile: 

I will remove the timer altogether then